transaction.go

```diff
@@ -49,198 +90,303 @@ type Transaction struct {
+	for i := uint64(0); i < requiredPages; i++ {
+		// Set offset according to the index in reservedPages
+		pages[i].offset = reservedPages[i]
+
+		// Set nextPage if the current page isn't the last one
+		// otherwise let it be nil
+		if i+1 < requiredPages {
+			pages[i].nextPage = &pages[i+1]
+		}
+
+		// Set pageStatus of the first page to pageStatusWritten
+		if i == 0 {
+			pages[i].pageStatus = pageStatusWritten
+		} else {
+			pages[i].pageStatus = pageStatusOther
+		}
+
+		// Copy part of the update into the payload
+		payloadsize := maxPayloadSize
+		if len(data[i*maxPayloadSize:]) < payloadsize {
+			payloadsize = len(data[i*maxPayloadSize:])
+		}
+		pages[i].payload = make([]byte, payloadsize)
+		copy(pages[i].payload, data[i*maxPayloadSize:])
+	}
+
+	// Set the first and final page of the transaction
+	t.firstPage = &pages[0]
+	t.finalPage = &pages[len(pages)-1]
+
+	// write the pages to disk and set the pageStatus
 	page := t.firstPage
-	var data []byte
 	for page != nil {
-		// make sure that the checksum field is just zeros before calculating the checksum
-		copy(page.transactionChecksum[:], make([]byte, crypto.HashSize))
-
-		bytes, err := page.Marshal()
+		err := page.Write(t.wal.logFile)
 		if err != nil {
-			return [crypto.HashSize]byte{}, build.ExtendErr("Failed to marshall page", err)
+			t.initComplete <- build.ExtendErr("Couldn't write the page to file", err)
+			return
 		}
-		data = append(data, bytes...)
 		page = page.nextPage
 	}
-	return crypto.HashBytes(data), nil
 }
 
 // validateChecksum checks if a transaction has been corrupted by computing a hash
-// and comparing it to the one in the finalPage of the transaction
+// and comparing it to the one in the firstPage of the transaction
 func (t Transaction) validateChecksum() error {
 	// Check if finalPage is set
-	if t.finalPage == nil {
-		return errors.New("Couldn't verify checksum. finalPage is nil")
+	if t.firstPage == nil {
+		return errors.New("Couldn't verify checksum. firstPage is nil")
 	}
 	checksum, err := t.checksum()
 	if err != nil {
 		return errors.New("Failed to create checksum for validation")
 	}
-	if checksum != t.finalPage.transactionChecksum {
+	if checksum != t.firstPage.transactionChecksum {
 		return errors.New("checksum not valid")
 	}
 	return nil
 }
 
-//commit commits a transaction by setting the correct status and checksum
-func (t *Transaction) commit() error {
-	// set the status of the first page first
-	t.firstPage.pageStatus = pageStatusComitted
-
-	// calculate the checksum and write it to the last page
-	checksum, err := t.checksum()
-	if err != nil {
-		return build.ExtendErr("Unable to create checksum of transaction", err)
-	}
-	t.finalPage.transactionChecksum = checksum
-	err = t.finalPage.Write(t.wal.logFile)
-	if err != nil {
-		return build.ExtendErr("Writing the final page failed", err)
+// NewTransaction creates a transaction from a set of updates
+func (w *WAL) NewTransaction(updates []Update) *Transaction {
```

Comment:
idiomatically, this function would be at the bottom because it's on the WAL and not the transaction. Functions of structs are sorted in a file by the structs that they operate on